sway-fmt-v2/src/utils/expr/tests.rs

```diff
@@ -0,0 +1,117 @@
+    let mut parser = Parser::new(&token_stream, &mut errors);
+    let expression: Expr = Parse::parse(&mut parser).unwrap();
+
+    let mut buf = Default::default();
+    expression.format(&mut buf, &mut formatter).unwrap();
+
+    buf
+}
+
+macro_rules! fmt_test {
+    ($scope:ident $desired_output:expr, $($name:ident $y:expr),+) => {
+        fmt_test_inner!($scope $desired_output,
+                                $($name $y)+
+                                ,
+                                remove_trailing_whitespace format!("{} \n\n\t ", $desired_output).as_str(),
+                                remove_beginning_whitespace format!("  \n\t{}", $desired_output).as_str(),
+                                identity $desired_output, /* test return is valid */
+                                remove_beginning_and_trailing_whitespace format!("  \n\t  {} \n\t   ", $desired_output).as_str()
+                       );
+    };
+}
+
+macro_rules! fmt_test_inner {
+    ($scope:ident $desired_output:expr, $($name:ident $y:expr),+) => {
+        $(
+        paste! {
+            #[test]
+            fn [<$scope _ $name>] () {
+                let formatted_code = format_code($y);
+                let changeset = diff_lines(&formatted_code, $desired_output);
+                let count_of_updates = changeset.diff().len();
+                if count_of_updates != 0 {
+                    println!("FAILED: {count_of_updates} diff items.");
+                }
+                for diff in changeset.diff() {
+                    match diff {
+                        DiffOp::Equal(old) => {
+                            for o in old {
+                                println!("{}", o)
+                            }
+                        }
+                        DiffOp::Insert(new) => {
+                            for n in new {
+                                println_green(&format!("+{}", n));
+                            }
+                        }
+                        DiffOp::Remove(old) => {
+                            for o in old {
+                                println_red(&format!("-{}", o));
+                            }
+                        }
+                        DiffOp::Replace(old, new) => {
+                            for o in old {
+                                println_red(&format!("-{}", o));
+                            }
+                            for n in new {
+                                println_green(&format!("+{}", n));
+                            }
+                        }
+                    }
+                }
+                assert_eq!(&formatted_code, $desired_output);
+            }
+        }
+    )+
+}
+}
+
+// TODO:
+// path, literal, abicast, struct, tuple, parens, block, array, asm, return, if, match, while, func app, index, field projection, tuple field projection, ref, deref, operators, bitwise stuff, comp operators
+// logical operators, reassignment
```

Comment:
Could we open an issue for these and link it here?

Reply:
sure! I have taken care of a few of them. I believe we would need all of them as tests before this PR can be approved.

Reply:
*at least the horizontal ones